### PR TITLE
[Feature] Support assign sort key of primary key model by schema change reorder.

### LINKED ISSUE #12684: [Feature] Support schema change primary key model sort key.
## Feature request

do not support assign sort key of primary key model by schema change reorder now.

**Is your feature request related to a problem? Please describe.**
```
MySQL [test]> CREATE TABLE test (
    ->   `k1` int,
    ->   `k2` int,
    ->   `k3` int,
    ->   `v1` int,
    ->   `v2` int,
    ->   `v3` int,
    ->   `v4` int,
    ->   `v5` int
    -> ) ENGINE=OLAP
    -> PRIMARY KEY(`k1`, `k2`, `k3`)
    -> COMMENT "OLAP"
    -> DISTRIBUTED BY HASH(`k1`, `k2`, `k3`) BUCKETS 1
    -> ORDER BY(`v2`, `v3`)
    -> PROPERTIES ("replication_num" = "1");
Query OK, 0 rows affected (0.11 sec)

MySQL [test]> ALTER TABLE test ORDER BY (v3, v2);
Primary key table do not support reorder column

MySQL [test]>
```

**Describe the solution you'd like**
```
MySQL [test]> select * from test;
+------+------+------+------+------+------+------+------+
| k1   | k2   | k3   | v1   | v2   | v3   | v4   | v5   |
+------+------+------+------+------+------+------+------+
|    1 |    5 |    1 |    5 |    1 |    5 |    1 |    5 |
|    2 |    4 |    2 |    4 |    2 |    4 |    2 |    4 |
|    3 |    3 |    3 |    3 |    3 |    3 |    3 |    3 |
|    4 |    2 |    4 |    2 |    4 |    2 |    4 |    2 |
|    5 |    1 |    5 |    1 |    5 |    1 |    5 |    1 |
+------+------+------+------+------+------+------+------+
5 rows in set (0.02 sec)

MySQL [test]> ALTER TABLE test ORDER BY (v3, v2);
Query OK, 0 rows affected (0.03 sec)

MySQL [test]> select * from test;
+------+------+------+------+------+------+------+------+
| k1   | k2   | k3   | v1   | v2   | v3   | v4   | v5   |
+------+------+------+------+------+------+------+------+
|    5 |    1 |    5 |    1 |    5 |    1 |    5 |    1 |
|    4 |    2 |    4 |    2 |    4 |    2 |    4 |    2 |
|    3 |    3 |    3 |    3 |    3 |    3 |    3 |    3 |
|    2 |    4 |    2 |    4 |    2 |    4 |    2 |    4 |
|    1 |    5 |    1 |    5 |    1 |    5 |    1 |    5 |
+------+------+------+------+------+------+------+------+
5 rows in set (0.02 sec)

MySQL [test]>
```

**Describe alternatives you've considered**
add reorder by sort key in schema change procedure.

### PR DESCRIPTION
**Pull request by Linkerist**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：
<!--
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #12684 #11733

## Problem Summary(Required) ：
<!-- (Please describe the changes you have made. In which scenarios will this bug be triggered and what measures have you taken to fix the bug?) -->

do not support assign sort key of primary key model by schema change reorder now:
```
MySQL [test]> CREATE TABLE test (
    ->   `k1` int,
    ->   `k2` int,
    ->   `k3` int,
    ->   `v1` int,
    ->   `v2` int,
    ->   `v3` int,
    ->   `v4` int,
    ->   `v5` int
    -> ) ENGINE=OLAP
    -> PRIMARY KEY(`k1`, `k2`, `k3`)
    -> COMMENT "OLAP"
    -> DISTRIBUTED BY HASH(`k1`, `k2`, `k3`) BUCKETS 1
    -> ORDER BY(`v2`, `v3`)
    -> PROPERTIES ("replication_num" = "1");
Query OK, 0 rows affected (0.11 sec)

MySQL [test]> ALTER TABLE test ORDER BY (v3, v2);
Primary key table do not support reorder column
```
should be:
```
MySQL [test]> select v2, v3 from test;
+------+------+
| v2   | v3   |
+------+------+
|    1 |    5 |
|    2 |    4 |
|    3 |    3 |
|    4 |    2 |
|    5 |    1 |
+------+------+
5 rows in set (0.02 sec)

MySQL [test]> show create table test;
+-------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                |
+-------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| test  | CREATE TABLE `test` (
  `k1` int(11) NOT NULL COMMENT "",
  `k2` int(11) NOT NULL COMMENT "",
  `k3` int(11) NOT NULL COMMENT "",
  `v1` int(11) NULL COMMENT "",
  `v2` int(11) NULL COMMENT "",
  `v3` int(11) NULL COMMENT "",
  `v4` int(11) NULL COMMENT "",
  `v5` int(11) NULL COMMENT ""
) ENGINE=OLAP 
PRIMARY KEY(`k1`, `k2`, `k3`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`k1`, `k2`, `k3`) BUCKETS 1 
ORDER BY(`v2`, `v3`)
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"storage_format" = "DEFAULT",
"enable_persistent_index" = "false",
"compression" = "LZ4"
); |
+-------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.01 sec)

MySQL [test]> ALTER TABLE test ORDER BY (v3, v2);
Query OK, 0 rows affected (0.03 sec)

MySQL [test]> select v3, v2 from test;
+------+------+
| v3   | v2   |
+------+------+
|    1 |    5 |
|    2 |    4 |
|    3 |    3 |
|    4 |    2 |
|    5 |    1 |
+------+------+
5 rows in set (0.01 sec)

MySQL [test]> show create table test;
+-------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                |
+-------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| test  | CREATE TABLE `test` (
  `k1` int(11) NOT NULL COMMENT "",
  `k2` int(11) NOT NULL COMMENT "",
  `k3` int(11) NOT NULL COMMENT "",
  `v1` int(11) NULL COMMENT "",
  `v2` int(11) NULL COMMENT "",
  `v3` int(11) NULL COMMENT "",
  `v4` int(11) NULL COMMENT "",
  `v5` int(11) NULL COMMENT ""
) ENGINE=OLAP 
PRIMARY KEY(`k1`, `k2`, `k3`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`k1`, `k2`, `k3`) BUCKETS 1 
ORDER BY(`v3`, `v2`)
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"storage_format" = "DEFAULT",
"enable_persistent_index" = "false",
"compression" = "LZ4"
); |
+-------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.01 sec)
```

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [x] This pr will affect users' behaviors
- [x] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 4 / 44 (09.09%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/alter/AlterJobV2Builder.java|0|2|00.00%|[88, 89]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeHandler.java|0|37|00.00%|[505, 507, 509, 510, 511, 512, 514, 515, 517, 518, 519, 520, 1031, 1032, 1036, 1039, 1040, 1041, 1042, 1043, 1044, 1046, 1047, 1048, 1050, 1051, 1052, 1054, 1057, 1058, 1060, 1061, 1063, 1209, 1210, 1211, 1214]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeJobV2.java|3|4|75.00%|[315]|
|:large_blue_circle:|com/starrocks/alter/OlapTableAlterJobV2Builder.java|1|1|100.00%|[]|

File: be/src/bench/CMakeLists.txt [hunk 1]
```diff
@@ -1,4 +1,4 @@
 # This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Inc.
 
-ADD_BE_BENCH(${SRC_DIR}/bench/chunks_sorter_bench)
-ADD_BE_BENCH(${SRC_DIR}/bench/runtime_filter_bench)
+#ADD_BE_BENCH(${SRC_DIR}/bench/chunks_sorter_bench)
```
Comment (chaoyli) [inline be/src/bench/CMakeLists.txt:3]: Why to annotate the two lines?

File: be/src/storage/rowset/rowset_writer.cpp [hunk 2]
```diff
@@ -573,15 +573,12 @@ Status HorizontalRowsetWriter::_final_merge() {
         ChunkIteratorPtr itr;
         // create temporary segment files at first, then merge them and create final segment files if schema change with sorting
         if (_context.schema_change_sorting) {
-            if (_context.tablet_schema->keys_type() == KeysType::DUP_KEYS) {
+            if (_context.tablet_schema->keys_type() == KeysType::DUP_KEYS ||
```
Comment (chaoyli) [inline be/src/storage/rowset/rowset_writer.cpp:576]: You have to describe the problem and solution clearly in the commit message.

File: be/src/bench/CMakeLists.txt (hunk 1 above)
Comment (Linkerist) [reply be/src/bench/CMakeLists.txt:3]: ok, changed. comment can reduce compile time, i forgot to uncomment it when push commit.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 4 / 44 (09.09%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/alter/AlterJobV2Builder.java|0|2|00.00%|[88, 89]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeHandler.java|0|37|00.00%|[505, 507, 509, 510, 511, 512, 514, 515, 517, 518, 519, 520, 1031, 1032, 1036, 1039, 1040, 1041, 1042, 1043, 1044, 1046, 1047, 1048, 1050, 1051, 1052, 1054, 1057, 1058, 1060, 1061, 1063, 1209, 1210, 1211, 1214]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeJobV2.java|3|4|75.00%|[315]|
|:large_blue_circle:|com/starrocks/alter/OlapTableAlterJobV2Builder.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 4 / 44 (09.09%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/alter/AlterJobV2Builder.java|0|2|00.00%|[88, 89]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeHandler.java|0|37|00.00%|[505, 507, 509, 510, 511, 512, 514, 515, 517, 518, 519, 520, 1031, 1032, 1036, 1039, 1040, 1041, 1042, 1043, 1044, 1046, 1047, 1048, 1050, 1051, 1052, 1054, 1057, 1058, 1060, 1061, 1063, 1209, 1210, 1211, 1214]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeJobV2.java|3|4|75.00%|[315]|
|:large_blue_circle:|com/starrocks/alter/OlapTableAlterJobV2Builder.java|1|1|100.00%|[]|

File: be/src/storage/rowset/rowset_writer.cpp [hunk 3]
```diff
@@ -670,15 +667,12 @@ Status HorizontalRowsetWriter::_final_merge() {
...
+                    _context.tablet_schema->keys_type() == KeysType::PRIMARY_KEYS) {
                     itr = new_mask_merge_iterator(seg_iterators, mask_buffer.get());
                 } else if (_context.tablet_schema->keys_type() == KeysType::UNIQUE_KEYS ||
                            _context.tablet_schema->keys_type() == KeysType::AGG_KEYS) {
                     itr = new_aggregate_iterator(new_mask_merge_iterator(seg_iterators, mask_buffer.get()), false);
-                } else {
-                    return Status::NotSupported(
-                            fmt::format("final merge: schema change with sorting do not support {} type",
-                                        KeysType_Name(_context.tablet_schema->keys_type())));
                 }
```
Comment (decster) [inline be/src/storage/rowset/rowset_writer.cpp:676]: why remove this, it will cause trouble if we add a new key type? if it's unreachable, add a assert or CHECK instead

File: be/test/exprs/vectorized/math_functions_test.cpp [hunk 4]
```diff
@@ -1302,8 +1302,8 @@ TEST_F(VecMathFunctionsTest, TrigonometricFunctionTest) {
     }
 
     {
-        std::vector<double> result_expect =
-            {std::tanh(-1), std::tanh(0), std::tanh(1), std::tanh(3.1415926), std::tanh(30)};
+        std::vector<double> result_expect = {std::tanh(-1), std::tanh(0), std::tanh(1), std::tanh(3.1415926),
+                                             std::tanh(30)};
         std::unique_ptr<FunctionContext> ctx(FunctionContext::create_test_context());
         ColumnPtr result = MathFunctions::tanh(ctx.get(), columns);
         auto v = ColumnHelper::cast_to<TYPE_DOUBLE>(result);
```
Comment (decster) [inline be/test/exprs/vectorized/math_functions_test.cpp:1309]: remove these unrelated change

File: fe/fe-core/src/main/java/com/starrocks/alter/SchemaChangeHandler.java [hunk 5]
```diff
@@ -1150,9 +1206,13 @@ public AlterJobV2 analyzeAndCreateJob(List<AlterClause> alterClauses, Database d
             } else if (alterClause instanceof ReorderColumnsClause) {
                 // reorder column
                 if (olapTable.getKeysType() == KeysType.PRIMARY_KEYS) {
-                    throw new DdlException("Primary key table do not support reorder column");
+                    List<Integer> sortKeyIdxes = new ArrayList<>();
+                    processReorderColumnOfPrimaryKey((ReorderColumnsClause) alterClause, olapTable, indexSchemaMap, sortKeyIdxes);
+                    return createJobForProcessReorderColumnOfPrimaryKey(db.getId(), olapTable,
+                            indexSchemaMap, sortKeyIdxes);
```
Comment (decster) [inline fe/fe-core/src/main/java/com/starrocks/alter/SchemaChangeHandler.java:1212]: what if a user made a mistake and use the old usage of `alter table order by`, for example:

```
create table table0 (a,b,c,d,e) primary key(a), order by(c,d) 

alter table order by (d, c, b, c, e)
```
Could we provide a notice of potential misusage if the user provides all the columns in the table in `alter table order by` for primary key table?

File: be/test/exprs/vectorized/math_functions_test.cpp (hunk 4 above)
Comment (Linkerist) [reply be/test/exprs/vectorized/math_functions_test.cpp]: ok, this change due to run clang-format.sh, can be restored since it none of this PR's business, changed.

File: fe/fe-core/src/main/java/com/starrocks/alter/SchemaChangeHandler.java (hunk 5 above)
Comment (Linkerist) [reply fe/fe-core/src/main/java/com/starrocks/alter/SchemaChangeHandler.java:1212]: old usage 1. need all columns. 2. all keys need be the prefix of all columns. 3. column can not be duplicate appear.
if write all columns, we can not tell the differences between make all columns to be sort key on purpose and making a mistake to use the old usage.
maybe the furthest the judgement can do is provide a notice when satisfy 1, 2, since that in the new usage 2 is needless.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 4 / 44 (09.09%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/alter/AlterJobV2Builder.java|0|2|00.00%|[88, 89]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeHandler.java|0|37|00.00%|[505, 507, 509, 510, 511, 512, 514, 515, 517, 518, 519, 520, 1031, 1032, 1036, 1039, 1040, 1041, 1042, 1043, 1044, 1046, 1047, 1048, 1050, 1051, 1052, 1054, 1057, 1058, 1060, 1061, 1063, 1209, 1210, 1211, 1214]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeJobV2.java|3|4|75.00%|[315]|
|:large_blue_circle:|com/starrocks/alter/OlapTableAlterJobV2Builder.java|1|1|100.00%|[]|

**Comment by Linkerist** [comment]
run starrocks_be_unittest

**Review by decster — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 13 / 43 (30.23%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeHandler.java|9|35|25.71%|[504, 505, 507, 508, 509, 510, 513, 514, 518, 520, 1044, 1045, 1049, 1052, 1053, 1054, 1055, 1056, 1057, 1059, 1060, 1061, 1063, 1064, 1065, 1212]|
|:large_blue_circle:|com/starrocks/alter/SchemaChangeJobV2.java|2|6|33.33%|[201, 314, 315, 316]|
|:large_blue_circle:|com/starrocks/alter/AlterJobV2Builder.java|2|2|100.00%|[]|

**Review by decster — APPROVED** [review]

File: be/src/storage/rowset/rowset_writer.cpp (hunk 2 above)
Comment (Linkerist) [reply be/src/storage/rowset/rowset_writer.cpp:576]: added.

**Review by chaoyli — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by Linkerist** [comment]
@Mergifyio backport branch-2.5
